masonry/src/tests/event.rs

```diff
@@ -153,4 +153,100 @@ fn click_anchors_focus() {
 
 // TEXT EVENTS
 
+#[test]
+fn text_event() {
+    let target_tag = WidgetTag::new("target");
+
+    let target = NewWidget::new_with_tag(TextArea::new_editable("").record(), target_tag);
+
+    let mut harness = TestHarness::create(default_property_set(), target);
+    let target_id = harness.get_widget_with_tag(target_tag).id();
+    harness.flush_records_of(target_tag);
+    harness.keyboard_type_chars("A");
+    harness.flush_records_of(target_tag);
+
+    // The widget isn't focused, it doesn't get text events.
+    harness.keyboard_type_chars("A");
+    assert_matches!(harness.get_records_of(target_tag)[..], []);
+
+    // We focus on the widget, now it gets text events.
+    harness.focus_on(Some(target_id));
+    harness.keyboard_type_chars("A");
+    let records = harness.get_records_of(target_tag);
+    assert!(records.iter().any(|r| matches!(r, Record::TextEvent(_))));
+}
+
+#[test]
+fn text_event_fallback() {
+    let target_tag = WidgetTag::new("target");
+    let parent_tag = WidgetTag::new("parent");
+
+    let child = NewWidget::new_with_tag(TextArea::new_editable("").record(), target_tag);
+    let parent = NewWidget::new_with_tag(Flex::row().with_child(child), parent_tag);
+
+    let mut harness = TestHarness::create(default_property_set(), parent);
+    harness.flush_records_of(target_tag);
+
+    // If the root widget has exactly one child, that child gets text events when no widget is focused.
+    harness.keyboard_type_chars("A");
+    let records = harness.get_records_of(target_tag);
+    assert!(records.iter().any(|r| matches!(r, Record::TextEvent(_))));
+
+    harness.edit_widget_with_tag(parent_tag, |mut flex| {
+        Flex::add_child(&mut flex, SizedBox::empty().with_auto_id());
+    });
+    harness.flush_records_of(target_tag);
+
+    // We've added another child, now nobody gets text events when no widget is focused.
+    harness.keyboard_type_chars("A");
+    assert_matches!(harness.get_records_of(target_tag)[..], []);
+}
+
+#[test]
+fn tab_focus() {
```

Comment:
As noted in #1226, we probably should at least think about our behaviour with nested focusable widgets.

Reply:
I'll add those tests later. Most of them will be in the `update.rs` file.

(Also, it's kind of awkward to write tests for behavior I'm planning to change anyway.)